enclave_build/src/image.rs

```diff
@@ -0,0 +1,287 @@
+    schema_version: u32,
+    media_type: Option<String>,
+
+    /// A reference to the image config
+    config: Descriptor,
+    
+    /// Array describing the image layers
+    layers: Vec<Descriptor>,
+    annotations: Option<HashMap<String, String>>,
+}
+
+#[derive(
+    Clone,
+    Debug,
+    Default,
+    Serialize,
+    Deserialize,
+    Eq,
+    PartialEq,
+    Getters,
+    Setters,
+)]
+#[getset(get = "pub", set = "pub")]
+/// The image configuration data, containing metadata of an image such as date created, author,
+/// architecture, as well as runtime data such as bash commands and default arguments.
+pub struct ImageConfig {
+    created: Option<String>,
+
+    author: Option<String>,
+
+    architecture: Option<String>,
+
+    os: Option<String>,
+
+    #[serde(rename = "os.version")]
+    os_version: Option<String>,
+
+    #[serde(rename = "os.features")]
+    os_features: Option<Vec<String>>,
+
+    config: Option<Config>,
+
+    /// Same structure as 'config' field, but usually this contains more detailed data
+    container_config: Option<Config>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    variant: Option<String>,
+
+    rootfs: RootFs,
+
+    history: Vec<LayerHistory>,
+}
+
+#[derive(
+    Clone,
+    Debug,
+    Default,
+    Serialize,
+    Deserialize,
+    Eq,
+    PartialEq,
+    Getters,
+    Setters,
+)]
+#[serde(rename_all = "PascalCase")]
+#[getset(get = "pub", set = "pub")]
+/// The execution parameters which should be used when running a container from
+/// the image.
+pub struct Config {
+    #[serde(rename = "Hostname")]
+    host_name: Option<String>,
+
+    #[serde(rename = "Domainname")]
+    domain_name: Option<String>,
+
+    attach_stdin: Option<bool>,
+    attach_stdout: Option<bool>,
+    attach_stderr: Option<bool>,
+    tty: Option<bool>,
+    open_stdin: Option<bool>,
+    stdin_once: Option<bool>,
+
+    user: Option<String>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    exposed_ports: Option<Vec<String>>,
+
+    image: Option<String>,
+
+    env: Option<Vec<String>>,
+    entrypoint: Option<Vec<String>>,
+    cmd: Option<Vec<String>>,
+    volumes: Option<Vec<String>>,
+    working_dir: Option<String>,
+    on_build: Option<Vec<String>>,
+    labels: Option<HashMap<String, String>>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    stop_signal: Option<String>,
+}
+
+#[derive(
+    Default,
+    Clone,
+    Debug,
+    Serialize,
+    Deserialize,
+    Eq,
+    PartialEq,
+    Getters,
+    Setters,
+)]
+#[getset(get = "pub", set = "pub")]
+/// References the layer content addresses used by the image.
+pub struct RootFs {
+    #[serde(rename = "type")]
+    type_field: String,
+
+    /// An array of layer content digest hashes.
+    diff_ids: Vec<String>,
+}
+
+#[derive(
+    Clone, Debug,
+    Default,
+    Serialize,
+    Deserialize,
+    Eq,
+    PartialEq,
+    Getters,
+    Setters,
+)]
+ #[getset(get = "pub", set = "pub")]
```

Comment:
If we had code coverage (which we don't, but still), it would be impossible to get a decent percentage with all these setters/getters.
You would have to write unit tests that set/get each field. That's annoying and waste of code.

My recommendation would be to write setters/getters manually only for the things you actually need, or if you need all of the fields, then make them public (not the best coding pattern, but it's the lesser of two evils).